qip-0014.md

```diff
@@ -0,0 +1,170 @@
+```
+ QIP: 14
+ Layer: EVM
+ Title: CREATE + CREATE2 opcode
+ Author: alanorwick <alan+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0014
+ Status: Draft
+ Type: Informational
+ Created: 2024-07-31
+ License: BSD-2-Clause
+```
+
+
+## Abstract
+
+This QIP proposes modifications to the CREATE opcode in Quai Network to ensure shard-specific address generation, while maintaining CREATE2 functionality as is. The proposal aims to adapt Ethereum's address creation mechanism to Quai's sharded architecture.
+
+## Motivation
+
+Quai Network's sharded architecture requires addresses to be shard-specific. The current Ethereum-style CREATE opcode does not guarantee this, potentially leading to cross-shard conflicts and inefficiencies. This proposal seeks to address this issue while maintaining as much compatibility with Ethereum as possible.
+
+## Specification
+
+### CREATE Opcode Modification
+
+1. The CREATE opcode will generate an address using the following method:
+   
+   a. Split the bytecode_data into two parts:
+      - `hashCode`: All but the last 4 bytes of bytecode_data
```

Comment:
What if a user provides a valid contract but no salt?

Reply:
then it is not a valid contract, according to this spec

Reply:
then it is not a valid contract, according to this spec

Reply:
It would be a valid contract and then result in a grinded address from the EVM.

Reply:
if they did not provide salt? the last 4 bytes of the data must always be interpreted as salt (otherwise ambiguity can lead to insecure edge cases). In that case, anyone that tries to interact with a contract, will ignore the last 4 bytes when they interpret the contract code.

Anyone that did not provide salt at last 4 bytes, will probably have a problem when the last 4 bytes are ignored (presumably initializer data)

Reply:
Correct, the last 4 bytes of data would also be interpreted as a salt. A contract is always valid with appended extra data.

Reply:
It's impossible to know without EVM execution whether the 4 bytes at the end are actual bytecode or appended extra data.